include/openthread/srp_client_service_pool.h

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * This function gets the array of IPv6 address entries to use as SRP client host address list.
+ *
+ * @param[in] aInstance      A pointer to the OpenThread instance.
+ * @param[out] aArrayLength  If not NULL, returns the array length (number of IPv6 address entries in the array).
+ *
+ * @returns A pointer to an array of `otIp6Address` entries (number of entries is returned in @p aArrayLength).
+ *
+ */
+otIp6Address *otSrpClientPoolGetHostAddressesArray(otInstance *aInstance, uint8_t *aArrayLength);
+
+/**
+ * This function allocates a new service entry from the pool.
+ *
+ * The returned service entry instance will be initialized as follows:
+ *
+ *  - `mService.mName` will point to an allocated string buffer which can be retrieved using the function
+ *                     `otSrpClientPoolGetServiceEntryServiceNameString()`.
+ *  - `mService.mInstanceName` will point to an allocated string buffer which can be retrieved using the function
+ *                     `otSrpClientPoolGetServiceEntryInstanceNameString()`.
+ *  - `mService.mTxtEntries` will point to `mTxtEntry`.
+ *  - `mService.mNumTxtEntries` will be set to one.
+ *  - Other `mService` fields (port, priority, weight) are set to zero.
+ *  - `mTxtEntry.mKey` is set to NULL (value is treated as already encoded).
+ *  - `mTxtEntry.mValue` will point to an allocated buffer which can be retrieved using the function
+ *                     `otSrpClientPoolGetServiceEntryTxtBuffer()`.
+ *  - mTxtEntry.mValueLength` is set to zero.
+ *
+ * @param[in] aInstance   A pointer to the OpenThread instance.
+ *
+ * @returns A pointer to the newly allocated service entry or NULL if not more entry available in the pool.
+ *
+ */
+otSrpClientPoolServiceEntry *otSrpClientPoolAllocateService(otInstance *aInstance);
+
+/**
+ * This function frees a previously allocated service entry.
+ *
+ * The @p aService MUST be previously allocated using `otSrpClientPoolAllocateService()` and not yet freed. Otherwise
+ * the behavior of this function is undefined.
+ *
+ * @param[in] aInstance   A pointer to the OpenThread instance.
+ * @param[in] aSevice     A pointer to the service entry to free (MUST NOT be NULL).
+ *
+ */
+void otSrpClientPoolFreeService(otInstance *aInstance, otSrpClientPoolServiceEntry *aService);
+
+/**
+ * This function frees all previously allocated service entries.
+ *
+ * @param[in] aInstance   A pointer to the OpenThread instance.
+ *
+ */
+void otSrpClientPoolFreeAllServices(otInstance *aInstance);
+
+/**
+ * This function gets the string buffer for service name from a service entry.
+ *
+ * @param[in]  aEntry   A pointer to a previously allocated service entry (MUST NOT be NULL).
+ * @param[out] aSize    If not NULL, returns the size (number of bytes) of the string buffer.
+ *
+ * @returns A pointer to the string buffer.
+ *
+ */
+char *otSrpClientPoolGetServiceEntryServiceNameString(otSrpClientPoolServiceEntry *aEntry, uint16_t *aSize);
```

Comment:
Why do we need an API to return the string buffer? Can't the client access the pointer directly in the struct? I understand the need to query the size.

Reply:
Yes, the caller can access the pointer to the `mName` and `mInstanceName` in `otSrpClientService` struct, but the challenge is that the they are defined as `const char *`, so the caller technically should not change the string buffer content. 

We can document it that caller needs to do `const_cast` but since we need a function to query the size of the buffer, I opted to return the pointer to the string buffer (now as `char *`) as well.